main.go

```diff
@@ -149,6 +161,33 @@ func (c *Controller) handleNewCSR(key string) error {
 	return nil
 }
 
+func maxPendingTotal(csrs []certificatesv1beta1.CertificateSigningRequest, machines []machinev1beta1.Machine, maxPendingCSRs int) int {
```

Comment:
I would expect:

`100 + 2*machines with no node ref + 1*machines with node ref`

i.e. `100 (buffer) + client+serving for new nodes + just serving for old nodes`

Reply:
2*(machine without node ref) doesn't fit because the second CSR is only generated after the first is approved.

I don't recall what the certificate rotation (if any) flow is.  That might generate both at once which would break us here.  There's also an edge case that if the csr approver is scaled down for some non-trivial amount of time, the kubelet will endless request CSRs every so often, but I'm not sure there's much we can do about this case in particular here.

All in all, I think 100 + 2*machines should probably be the logic in case there's a large inflow of renewals that need signing, there shouldn't be more than 2 CSRs per node.

Reply:
How about a slightly different approach:

1. Parse CSR
1. If CSR is not for node client cert, do no max pending check (meaning we do not care about having a million pending serving cert CSRs at once)
1. If CSR is for node client cert
    1. check to see how many machines have missing node refs, add `10`, store as `maxPending`
    1. check to see how many node client certs CSRs are pending, must not be greater than `maxPending`

Reply:
@enj given this is mostly a bug fix and I'm assuming we want to backport to 4.1, I'd prefer to keep it simple for now.

Reply:
> @enj given this is mostly a bug fix and I'm assuming we want to backport to 4.1, I'd prefer to keep it simple for now.

The approach in master is simple and has proven to be incorrect.  I find this to also be incorrect, just less so.  Please take the time to do it correctly.  We are not rushing to get this out the door like 4.1.

Reply:
@enj I did take the time to do it correctly.  If you'll notice, the original suggestion on how to implement it was wrong.  I don't want to add more complexity to this, if you want to do it another way, feel free to implement it.